lib/Dialect/Comb/InferIntRangeInterfaceImpls.cpp

```diff
@@ -0,0 +1,316 @@
+       totalWidth += operandWidth) {
+    auto uminUpd = uminIn.ushl_sat(totalWidth);
+    auto umaxUpd = umaxIn.ushl_sat(totalWidth);
+    umin = umin.uadd_sat(uminUpd);
+    umax = umax.uadd_sat(umaxUpd);
+  }
+  auto urange = ConstantIntRanges::fromUnsigned(umin, umax);
+  setResultRange(getResult(), urange);
+};
+
+//===----------------------------------------------------------------------===//
+// MuxOp
+//===----------------------------------------------------------------------===//
+
+void comb::MuxOp::inferResultRangesFromOptional(
+    ArrayRef<IntegerValueRange> argRanges, SetIntLatticeFn setResultRange) {
+  std::optional<APInt> mbCondVal =
+      argRanges[0].isUninitialized()
+          ? std::nullopt
+          : argRanges[0].getValue().getConstantValue();
+
+  const IntegerValueRange &trueCase = argRanges[1];
+  const IntegerValueRange &falseCase = argRanges[2];
+
+  if (mbCondVal) {
+    if (mbCondVal->isZero())
+      setResultRange(getResult(), falseCase);
+    else
+      setResultRange(getResult(), trueCase);
+    return;
+  }
+  setResultRange(getResult(), IntegerValueRange::join(trueCase, falseCase));
+}
+
+//===----------------------------------------------------------------------===//
+// ICmpOp
+//===----------------------------------------------------------------------===//
+
+void comb::ICmpOp::inferResultRanges(ArrayRef<ConstantIntRanges> argRanges,
+                                     SetIntRangeFn setResultRange) {
+  comb::ICmpPredicate combPred = getPredicate();
+
+  APInt min = APInt::getZero(1);
+  APInt max = APInt::getAllOnes(1);
+
+  if (combPred == comb::ICmpPredicate::ceq ||
+      combPred == comb::ICmpPredicate::cne ||
+      combPred == comb::ICmpPredicate::weq ||
+      combPred == comb::ICmpPredicate::wne) {
+    // These predicates are not supported for integer range analysis
+    setResultRange(getResult(), ConstantIntRanges::fromUnsigned(min, max));
+    return;
+  }
+
+  intrange::CmpPredicate pred;
+  switch (combPred) {
+  case comb::ICmpPredicate::eq:
+    pred = intrange::CmpPredicate::eq;
+    break;
+  case comb::ICmpPredicate::ne:
+    pred = intrange::CmpPredicate::ne;
+    break;
+  case comb::ICmpPredicate::slt:
+    pred = intrange::CmpPredicate::slt;
+    break;
+  case comb::ICmpPredicate::sle:
+    pred = intrange::CmpPredicate::sle;
+    break;
+  case comb::ICmpPredicate::sgt:
+    pred = intrange::CmpPredicate::sgt;
+    break;
+  case comb::ICmpPredicate::sge:
+    pred = intrange::CmpPredicate::sge;
+    break;
+  case comb::ICmpPredicate::ult:
+    pred = intrange::CmpPredicate::ult;
+    break;
+  case comb::ICmpPredicate::ule:
+    pred = intrange::CmpPredicate::ule;
+    break;
+  case comb::ICmpPredicate::ugt:
+    pred = intrange::CmpPredicate::ugt;
+    break;
+  case comb::ICmpPredicate::uge:
+    pred = intrange::CmpPredicate::uge;
+    break;
+  default:
+    llvm_unreachable("Unknown comparison predicate");
+  }
```

Comment:
nit: Could you return in default? 
```suggestion
  
  intrange::CmpPredicate pred;
  switch (combPred) {
  case comb::ICmpPredicate::eq:
    pred = intrange::CmpPredicate::eq;
    break;
  case comb::ICmpPredicate::ne:
    pred = intrange::CmpPredicate::ne;
    break;
  case comb::ICmpPredicate::slt:
    pred = intrange::CmpPredicate::slt;
    break;
  case comb::ICmpPredicate::sle:
    pred = intrange::CmpPredicate::sle;
    break;
  case comb::ICmpPredicate::sgt:
    pred = intrange::CmpPredicate::sgt;
    break;
  case comb::ICmpPredicate::sge:
    pred = intrange::CmpPredicate::sge;
    break;
  case comb::ICmpPredicate::ult:
    pred = intrange::CmpPredicate::ult;
    break;
  case comb::ICmpPredicate::ule:
    pred = intrange::CmpPredicate::ule;
    break;
  case comb::ICmpPredicate::ugt:
    pred = intrange::CmpPredicate::ugt;
    break;
  case comb::ICmpPredicate::uge:
    pred = intrange::CmpPredicate::uge;
    break;
  default:
    // These predicates are not supported for integer range analysis
    setResultRange(getResult(), ConstantIntRanges::fromUnsigned(min, max));
    return;
  }
```

Reply:
Resolved!